### DISCUSSION TITLE
FastAPI StreamingResponse start download while waiting for data

### DISCUSSION THREAD

**Discussion by valeriodeblasio-1** [fastapi/fastapi#11526] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [x] I commit to help with one of those options 👆

### Example Code

```python
async def async_read_file():
    # this code simulate the activate waiting for the file to be created
    count = 0
    while count < 5:
        await asyncio.sleep(2)
        count += 1

    file = open('file.zip', 'rb')
    while True:
        data = file.read(64 * 1024)
        if not data:
            break
        yield data


@app.get('/download')
async def download():
    async def iterate_file_():
        async for chunk in async_read_file():
            yield chunk

    response = StreamingResponse(iterate_file_(), headers={
        "Content-Disposition": "attachment; filename*=utf-8''{}".format(quote('file.zip'))})

    return response
```

### Description

I want to send a file to a user, however the file is generated asynchronously via a thread which can take a long time (even 30 seconds).

I would like the browser to show the file being downloaded right away (with 0 bytes and unknown size), tried via cUrl and it works but with a browser (Chrome or Firefox) this doesn't happen, the download only starts after the first bytes have been sent.

The only way I've been able to find is to send empty characters (\u200B) while waiting for the file to be generated (as suggested in this similar problem [Chrome doesn't show file as being downloaded until 8 bytes are sent (Firefox does)](https://stackoverflow.com/questions/52913925/chrome-doesnt-show-file-as-being-downloaded-until-8-bytes-are-sent-firefox-doe)), but this way the resulting file contains these characters at the beginning, which then gives problems when loading on other systems.

### Operating System

Linux, macOS

### FastAPI Version

0.109.2

### Pydantic Version

2.1.1

### Python Version

3.11.2

**Comment by patrick91** [comment]
@valeriodeblasio-1 you can do this with the `application/force-download` content type 😊

```python
@app.get("/download")
async def download():
    async def iterate_file_():
        async for chunk in async_read_file():
            yield chunk

    response = StreamingResponse(
        iterate_file_(),
        headers={
            "Content-Type": "application/force-download", # 👈
            "Content-Disposition": "attachment; filename*=utf-8''{}".format("file.zip"),
        },
    )

    return response
```

**Reply by valeriodeblasio-1** [reply]
Thanks man, it seems that “force-download” only works for Chrome, but not for Firefox (I've also read that it's not a standard MIME, so I don't know if it would break anything).

**Comment by YuriiMotov** [comment]
Let's close this for now to keep the discussions clean and organized.
If you’ve found a solution, please share it here.
If you still need help, feel free to let us know.
